src/core/ddsi/src/ddsi_sertype.c

```diff
@@ -118,40 +118,38 @@ struct sertype_ser
 
 const enum pserop sertype_ser_ops[] = { XS, Xb, XSTOP };
 
-size_t ddsi_sertype_serialize_size (const struct ddsi_sertype *tp)
+bool ddsi_sertype_serialize (const struct ddsi_sertype *tp, size_t *dst_sz, unsigned char **dst_buf)
 {
   struct sertype_ser d = { tp->type_name, tp->typekind_no_key };
-  size_t sz = 0;
-  plist_ser_generic_size_embeddable (&sz, &d, 0, sertype_ser_ops);
-  return sz;
-}
+  size_t dst_pos = 0;
 
-bool ddsi_sertype_serialize (const struct ddsi_sertype *tp, size_t *dst_pos, unsigned char *dst_buf)
-{
-  assert (dst_pos);
-  assert (dst_buf);
-  struct sertype_ser d = { tp->type_name, tp->typekind_no_key };
-  return (plist_ser_generic_embeddable ((char *) dst_buf, dst_pos, &d, 0, sertype_ser_ops, false) >= 0);
+  if (tp->ops->serialized_size == NULL || tp->ops->serialize == NULL)
+    return false;
+  *dst_sz = 0;
+  plist_ser_generic_size_embeddable (dst_sz, &d, 0, sertype_ser_ops);
+  tp->ops->serialized_size (tp, dst_sz);
+  *dst_buf = ddsrt_malloc (*dst_sz);
+  if (plist_ser_generic_embeddable ((char *) *dst_buf, &dst_pos, &d, 0, sertype_ser_ops, false) < 0)
+    return false;
+  if (!tp->ops->serialize (tp, &dst_pos, *dst_buf))
+    return false;
+  assert (dst_pos == *dst_sz);
+  return true;
 }
 
-bool ddsi_sertype_deserialize (struct ddsi_sertype *tp, size_t src_sz, const unsigned char *src_data, size_t *src_pos)
+bool ddsi_sertype_deserialize (struct ddsi_domaingv *gv, struct ddsi_sertype *tp, const struct ddsi_sertype_ops *sertype_ops, size_t sz, unsigned char *serdata)
 {
-  assert (src_pos);
   struct sertype_ser d;
   size_t srcoff = 0;
-  if (plist_deser_generic_srcoff (&d, src_data, src_sz, &srcoff, false, sertype_ser_ops) < 0)
+  if (plist_deser_generic_srcoff (&d, serdata, sz, &srcoff, DDSRT_ENDIAN != DDSRT_LITTLE_ENDIAN, sertype_ser_ops) < 0)
```

Comment:
The last parameter of `plist_ser_generic_embeddable` is a bool that indicates big-endian encoding, not bswap y/n, so that's set to `false` because the xtypes spec (7.3.4.5) requires LE encoding for TypeObject serialization. The deser functions have a `bool bswap` parameter, which make it somewhat confusing. So maybe we could change the ser functions to use `bool bswap` as well?

Reply:
> because the xtypes spec (7.3.4.5) requires LE encoding for TypeObject serialization

Dealing with OMG specs is like dealing with Windows: cruel and unusual punishment — but at least the organisation has a good acronym ... So the security spec mandates big-endian encoding (unnecessarily) while the XTypes spec mandates little-endian encoding (equally unnecessarily) even though the original DDSI specification leaves the decision to the sender except in the one case where a standard endianness is required (the perfidious keyhashes (\*)) and mandates that the receiver can handle either. (I admit to not having read all the fine print in the XTypes spec ...)

The last parameter of `plist_ser_generic_embeddable` is a bool that controls native vs big-endian encoding and was added to accomodate the Security spec's foibles, but setting it to `false` doesn't allow for the XTypes spec's insanity because on a big-endian machine that'll give you big-endian. So it needs to be complexified further and changed into a tri-state.

(\*) The keyhash is perfidious because for keys of which the serialized length may be larger than 16 bytes, it uses the MD5 hash, and therefore creating a topic with an octet sequence as key and writing two instances with keys that happen to produce an MD5 hash collision has "interesting" consequences for implementations that rely on the keyhash. And yet the DDS Security spec mandates the presence of the keyhash for those cases where the readers can't decode the data but need to keep instances separate, effectively forbidding applications from choosing the key values freely, but rather requiring them to avoid MD5 collisions. Not much good can be said about that.

Reply:
P.S.: of course a boolean that says whether or not to byteswap while serialising suffices to do any of the modes we need. But that pushes the tests like `DDSRT_ENDIAN != DDSRT_LITTLE_ENDIAN`) to the caller. I think the interface would be easier to use if you could specify the desired result as native/little/big endian and then let the serializer function figure out whether or not it needs to do byteswapping.